src/source/WasmGenerator.scala

```diff
@@ -365,13 +365,22 @@ class WasmGenerator(spec: Spec) extends Generator(spec) {
             s"${stubParamType(p.ty)} ${stubParamName(p.ident)}"
           }).mkString(","))
           w.w(")").braced {
-            if (!m.ret.isEmpty) w.w("auto r = ")
-            if (m.static) w.w(s"$cls::") else w.w("self->")
-            writeAlignedCall(w, s"""${idCpp.method(m.ident)}(""", m.params, ")", p => {
-              s"${helperClass(p.ty.resolved)}::toCpp(${stubParamName(p.ident)})"
-            })
-            w.wl(";")
-            m.ret.fold()(r => w.wl(s"return ${helperClass(r.resolved)}::fromCpp(${cppMarshal.maybeMove("r", r)});"))
+            w.w("try").braced {
+              if (!m.ret.isEmpty) w.w("auto r = ")
+              if (m.static) w.w(s"$cls::") else w.w("self->")
+              writeAlignedCall(w, s"""${idCpp.method(m.ident)}(""", m.params, ")", p => {
+                s"${helperClass(p.ty.resolved)}::toCpp(${stubParamName(p.ident)})"
+              })
+              w.wl(";")
+              m.ret.fold()(r => w.wl(s"return ${helperClass(r.resolved)}::fromCpp(${cppMarshal.maybeMove("r", r)});"))
+            }
+            w.w("catch(const std::exception& e)").braced {
```

Comment:
This seems reasonable for exceptions that were natively from C++. But for exceptions that originated in JS, can we re-create the original JS error (as best as we can)? We have this [JS exception](https://github.com/Snapchat/djinni/blob/11e3bc97ab59819da1254c67ce6fab1f3fa2bd27/support-lib/wasm/djinni_wasm.cpp#L76), which we can store more information in beyond just the message. We could read the relevant fields out of the JS one, hold them in the C++ one, then put them back in the JS one here when we re-throw

Reply:
good idea.  I think I'm going to try to store the JS Error object inside the C++ exception, so that later I can unwrap it and let JS code retrieve the original error object.

Reply:
That would be neat. Let's try it and see if it works. Preserving the correct stack trace as best as possible would be awesome. So that we're not losing information as the error bubbles up